### ISSUE TITLE
DL3006 fails when FROM uses an ARG

### ISSUE THREAD

**Issue by jhoblitt** [hadolint/hadolint#219]
- [x] This is a bug report

### Expected behavior

hadolint exits `0`

### Actual behavior

hadolint exists `1`

### Steps to reproduce the behavior

```bash
$ cat Dockerfile 
ARG BASE_IMAGE=alpine:3.7
FROM $BASE_IMAGE
$ docker run -ti -v$(pwd):$(pwd) -w $(pwd) hadolint/hadolint:v1.7.2 hadolint Dockerfile
Dockerfile:2 DL3006 Always tag the version of an image explicitly
```

**Comment by lorenzo** [comment]
out of curiosity, what is the benefit of doing this? Do you set from the command line a different base image?

**Comment by jhoblitt** [comment]
@lorenzo That's correct.  It is useful if you are producing several similar images, such as from both a centos 6 and centos 7 base image, under CI/CD.  Prior to docker gaining support for ARG in FROM, many users were resorting to templating only to change the `FROM`.

**Comment by lorenzo** [comment]
Do you often change the base image name instead of the tag? I imagine that all the instructions for that file would depend on something centros, for instance. 

Would it make sense if the longer suggested to use the arg as the tag part, instead of the full image?

**Comment by jhoblitt** [comment]
Yes, I change the docker registry repo name when testing.

**Comment by lorenzo** [comment]
I see... seems like we can just disable the rule for this type of use case. I wish there was a way we could still provide some safety, but I guess this is too custom to even try.

Will look into disabling the rule and release a new version soon

**Comment by jhoblitt** [comment]
I agree that the best option is to detect this case rather than fail with a false positive.  I suppose `FROM ${MYARG}:mystatictag` could still be checked for.

**Comment by yurrriq** [comment]
FWIW we do more like the following, and I'd like it if Hadolint didn't start to complain.

```docker
ARG MYTAG=x.y.z
FROM staticimage:${MY_TAG}
```

**Comment by zemanlx** [comment]
@yurrriq I am doing the same. The important thing is to have `:` in string after `FROM` and as we do not interpret variables we don't know if `:` is in a variable value. (Plus other rules that apply.)

**Comment by pakraticus** [comment]
Perhaps adding a FAQ to the wiki starting with

Q: How do I suppress a one line warning because hadolint cannot expand a variable yet and will never know what I intend to put in an arg?
A: Use the inline '# hadolint ignore=....'

Q: I want to know when to clean up my Dockerfile when hadolint can expand an environment variable.
Write a test Dockerfile IE
FROM alpine:3.7
ENV PACKAGE=gcc=3.5
RUN apk add --no-cache "${PACKAGE}"
And a test runner like
hadolint Dockerfile.test | grep DL3018

Back to the initial issue...  This approach seems to fit 'docker image list' output format.

```
ARG REPOSITORY
ARG TAG
FROM ${REPOSITORY}:${TAG}
```

Which came in handy when evaluating Jenkins and Gerrit against several Java images.
Next time I'd use a template language and run hadolint against the generated files.

**Comment by soullivaneuh** [comment]
I have the same issue using multi-layer to build many variant of my PHP image:

```dockerfile
ARG VARIANT=rc
ARG OS=debian

FROM registry.gitlab.com/nexylan/docker/core:1.12.1 as core
FROM composer:1.9.3 as composer

FROM php:${VARIANT} as build-debian
RUN apt-get update && apt-get install --yes --no-install-recommends unzip=6.* \
  && apt-get clean \
  && rm --recursive --force /var/lib/apt/lists/*

FROM php:${VARIANT}-alpine as build-alpine
RUN apk add --no-cache unzip~=6

FROM build-${OS} as build
```

Got:

```
./Dockerfile:15 DL3006 Always tag the version of an image explicitly
```

On that case, I don't see how I can use tags.

You can see the full codebase here: https://gitlab.com/nexylan/docker/php/-/merge_requests/16

**Comment by zemanlx** [comment]
@soullivaneuh In your case, I would use inline ignore

```Dockerfile
# hadolint ignore=DL3006
FROM build-${OS} as build
```

**Comment by guillaume-d** [comment]
This is a duplicate (subset) of #339 AFAICS.

**Comment by lorenzo** [comment]
closing in favor of #339
